### PR TITLE
HDFStore.walk() to iterate on groups

### LINKED ISSUE #10143: ENH: Iterate over HDF store hierarchically
I am relatively new to pandas, but have been using raw HDF5 files for a long itme, so apologies if this feature or something equivalent already exists.

As stated in the documentation, it is possible to create hierarchies of data stores:

``` python
store.put('/first_group/df1', df1)
store.put('/first_group/df2', df2)
store.put('/second_group/df3', df3)
```

However, it seems that `store.keys()` can only provide a flattened list of all available groups.  It would be nice to be able to walk the HDF5 file hierarchically.  A couple examples (this is borrowing a bit from the h5py API):

``` python
# do stuff on each subgroup/substore in /first_group
for s in store['first_group'].values() :
    # do stuff on df1 and df2

# visit each group at the root level
for s in store.values() :
    if 'df1' in s :
        # do stuff on any store named 'df1'
```

The main use case I would have for something like this is that I have many dataframes with different schema but would like to group some of them together and then iterate through those groups. 

Best way I can think to emulate this is to parse the strings in `store.keys()`, but that is pretty ugly.

### PR DESCRIPTION
**Pull request by CharlesB2**
This PR adds a walk() method to HDFStore in order to iterate on groups.

This is a revival of the initial PR #10932, rebased on upstream/master and updated tests.

- [x] closes #10143 
- [x] tests added / passed: test_pytables.py:testwalk()
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/pytables.py [hunk 1]
```diff
@@ -1106,6 +1106,46 @@ def groups(self):
...
+        pandas object names for each group.  Any non-pandas PyTables objects
+        that are not a group will be ignored.
+
+        The where group itself is listed first (preorder), then each of its
+        child groups (following an alphanumerical order) is also traversed,
+        following the same procedure.  If where is not supplied, the root
+        group is used.
+
+        The where argument can be a path string
+        or a Group instance (see :ref:`GroupClassDescr`).
```
Comment (jreback) [inline pandas/io/pytables.py]: what does this refer to? best to directly link to the pytables docs

File: pandas/io/pytables.py [hunk 2]
```diff
@@ -1106,6 +1106,46 @@ def groups(self):
                  g._v_name != u('table'))))
         ]
 
+    def walk(self, where="/"):
+        """ Walk the pytables group hierarchy yielding the group name and
```
Comment (jreback) [inline pandas/io/pytables.py]: should be a single sumary line, then a multi-line extended summary

File: pandas/io/pytables.py [hunk 3]
```diff
@@ -1106,6 +1106,46 @@ def groups(self):
...
+        pandas object names for each group.  Any non-pandas PyTables objects
+        that are not a group will be ignored.
```
Comment (jreback) [inline pandas/io/pytables.py]: need a Parameters section

File: pandas/tests/io/test_pytables.py [hunk 4]
```diff
@@ -4999,6 +4999,62 @@ def test_read_nokey_empty(self):
...
+            'df1': pd.DataFrame([1, 2, 3]),
+            'df2': pd.DataFrame([4, 5, 6]),
+            'df3': pd.DataFrame([6, 7, 8]),
+            'df4': pd.DataFrame([9, 10, 11]),
+            's1': pd.Series([10, 9, 8]),
+            'a1': np.array([[1, 2, 3], [4, 5, 6]])
+        }
+
+        with ensure_clean_store('walk_groups.hdf', mode='w') as store:
+            store.put('/first_group/df1', objs['df1'])
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: add a table or 2

File: pandas/tests/io/test_pytables.py [hunk 5]
```diff
@@ -4999,6 +4999,62 @@ def test_read_nokey_empty(self):
             store.close()
             pytest.raises(ValueError, read_hdf, path)
 
+    # GH10143
+    def test_walk(self):
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: put the comment inside the test function

File: pandas/tests/io/test_pytables.py [hunk 6]
```diff
@@ -4999,6 +4999,62 @@ def test_read_nokey_empty(self):
             store.close()
             pytest.raises(ValueError, read_hdf, path)
 
+    # GH10143
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: put this near tests for get_node

File: pandas/tests/io/test_pytables.py [hunk 7]
```diff
@@ -4999,6 +4999,62 @@ def test_read_nokey_empty(self):
...
+            store.put('/first_group/df2', objs['df2'])
+            store.put('/second_group/df3', objs['df3'])
+            store.put('/second_group/s1', objs['s1'])
+            store.put('/second_group/third_group/df4', objs['df4'])
+            g1 = store._handle.get_node('/first_group')
+            store._handle.create_array(g1, 'a1', objs['a1'])
+
+            expect1 = {
+                '': ({'first_group', 'second_group'}, set()),
+                '/first_group': (set(), {'df1', 'df2'}),
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: can you parameterize this test rather than making a long test like this

File: doc/source/io.rst [hunk 8]
```diff
@@ -3554,6 +3554,22 @@ everything in the sub-store and **below**, so be *careful*.
    store.remove('food')
    store
 
+
+You can walk through the group hierarchy using the ``walk`` method which
```
Comment (jreback) [inline doc/source/io.rst]: add a versionadded

File: pandas/io/pytables.py (hunk 2 above)
Comment (jreback) [inline pandas/io/pytables.py]: add a versionadded

**Review by jreback — CHANGES_REQUESTED** [review]
some comments

File: pandas/tests/io/test_pytables.py (hunk 4 above)
Comment (CharlesB2) [reply pandas/tests/io/test_pytables.py]: By table do you mean store one of the dataframes in table format?

File: pandas/tests/io/test_pytables.py (hunk 6 above)
Comment (CharlesB2) [reply pandas/tests/io/test_pytables.py]: Sorry, but I couldn't spot any test for get_node in the repo?

File: pandas/io/pytables.py (hunk 1 above)
Comment (CharlesB2) [reply pandas/io/pytables.py]: Actually it doesn't make sense for `where` to be anything else than `str`, removing this mention

File: doc/source/io.rst (hunk 8 above)
Comment (CharlesB2) [reply doc/source/io.rst]: done

File: pandas/tests/io/test_pytables.py (hunk 7 above)
Comment (CharlesB2) [reply pandas/tests/io/test_pytables.py]: done

File: pandas/tests/io/test_pytables.py (hunk 6 above)
Comment (CharlesB2) [reply pandas/tests/io/test_pytables.py]: moved

File: pandas/tests/io/test_pytables.py (hunk 5 above)
Comment (CharlesB2) [reply pandas/tests/io/test_pytables.py]: done

File: pandas/tests/io/test_pytables.py (hunk 4 above)
Comment (CharlesB2) [reply pandas/tests/io/test_pytables.py]: added

File: pandas/io/pytables.py (hunk 3 above)
Comment (CharlesB2) [reply pandas/io/pytables.py]: added

File: pandas/io/pytables.py (hunk 2 above)
Comment (CharlesB2) [reply pandas/io/pytables.py]: done

**Comment by CharlesB2** [comment]
I think I addressed your review's comments

File: doc/source/whatsnew/v0.24.0.txt [hunk 9]
```diff
@@ -16,7 +16,8 @@ Other Enhancements
 - :func:`Series.mode` and :func:`DataFrame.mode` now support the ``dropna`` parameter which can be used to specify whether NaN/NaT values should be considered (:issue:`17534`)
 - :func:`to_csv` now supports ``compression`` keyword when a file handle is passed. (:issue:`21227`)
 - :meth:`Index.droplevel` is now implemented also for flat indexes, for compatibility with MultiIndex (:issue:`21115`)
-
+- New method :meth:`HDFStore.walk` will recursively walk the group hierarchy of a HDF5 file (:issue:`10932`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you add this to api.rst as well

Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: an HDF5 file

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment. can you elaborate on why you think the return values are the right thing here? e.g. is this from the PyTables api directly? e.g. what is the point of having the groups and subkeys? wouldn't you just care about the subkeys?

**Comment by CharlesB2** [comment]
> minor comment. can you elaborate on why you think the return values are the right thing here? e.g. is this from the PyTables api directly? e.g. what is the point of having the groups and subkeys? wouldn't you just care about the subkeys?

It was made this way in the original PR, I didn't see any reason to change it. It's a refinement from the PyTables api, I guess the original author  @stephenpascoe  found it convenient. 

I made the docs change in the last forced push.

File: doc/source/whatsnew/v0.24.0.txt (hunk 9 above)
Comment (CharlesB2) [reply doc/source/whatsnew/v0.24.0.txt]: Done

File: doc/source/whatsnew/v0.24.0.txt [hunk 10]
```diff
@@ -17,8 +17,9 @@ Other Enhancements
 - :func:`to_datetime` now supports the ``%Z`` and ``%z`` directive when passed into ``format`` (:issue:`13486`)
 - :func:`Series.mode` and :func:`DataFrame.mode` now support the ``dropna`` parameter which can be used to specify whether NaN/NaT values should be considered (:issue:`17534`)
 - :func:`to_csv` now supports ``compression`` keyword when a file handle is passed. (:issue:`21227`)
-- :meth:`Index.droplevel` is now implemented also for flat indexes, for compatibility with :class:`MultiIndex` (:issue:`21115`)
-
+- :meth:`Index.droplevel` is now implemented also for flat indexes, for compatibility with MultiIndex (:issue:`21115`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: this was updated and the diff is taking the original, can you restore (the droplevel line)

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment. looks good. ping on green.

File: doc/source/whatsnew/v0.24.0.txt (hunk 10 above)
Comment (CharlesB2) [reply doc/source/whatsnew/v0.24.0.txt]: done

**Comment by jreback** [comment]
merged via https://github.com/pandas-dev/pandas/commit/45e55af9cf74aede00a8b10f8922537f285a573a

for some reason couldn't push to your remote, so manually merged.

thanks!

**Pull request closed without merging**